### PR TITLE
Add ExamsCard

### PR DESCRIPTION
**Pull request by BuildmodeOne**
image

This card is shown by default for students but is only visible if there are exams in the next 30 days. 
If there are not exams, the card is completely hidden. 

### Summary
🗓️📝🌐

Added a new feature to show upcoming exams from the `exams` platform on the dashboard and the calendar page. Implemented a new `ExamsCard` component to display the exams on the dashboard, and added a `focus` query parameter to the calendar page to select the default tab. Added translations for the new texts in German and English.

> _The dashboard has a new `ExamsCard`_
> _To show you the tests that are not hard_
> _It uses some hooks_
> _And translation books_
> _And links to the `calendar` with a guard_

### Walkthrough
*  Add a new dashboard card for upcoming exams ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-af4ec176af3bf8a0f9b885f1a9abeda79272eed848468a7375071c713d8304e6R9), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-af4ec176af3bf8a0f9b885f1a9abeda79272eed848468a7375071c713d8304e6R35-R40), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-bf718cc9bdccb5b738584020f42d1e204d73fd4dd68189f16694b350108fdecbR1-R79), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-fb32edcf9c6b453131b259e02d545d2c43b53a2e70850af7b1154800b25b5e9dL51-R56), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-34bd987125c20aa08a7cd7547b3e3f2bcfb581fd07ad4bbad8dcebe742da4acbL51-R56))
  * Import the `ExamsCard` component from `cards` folder in `allCards.js` ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-af4ec176af3bf8a0f9b885f1a9abeda79272eed848468a7375071c713d8304e6R9))
  * Register the `exams` object in the `ALL_DASHBOARD_CARDS` array in `allCards.js` with key, removable status, default platforms and user kinds, and card rendering function ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-af4ec176af3bf8a0f9b885f1a9abeda79272eed848468a7375071c713d8304e6R35-R40))
  * Implement the `ExamsCard` component in `cards/ExamsCard.js` using various hooks and modules to fetch, filter, and format the exam data, and render the card layout with a list group and a link to the calendar page ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-bf718cc9bdccb5b738584020f42d1e204d73fd4dd68189f16694b350108fdecbR1-R79))
  * Add the `exam` and `exams` keys and values to the German and English translation files for the `dashboard` namespace to provide the translation for the card title and the exam events prefix ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-fb32edcf9c6b453131b259e02d545d2c43b53a2e70850af7b1154800b25b5e9dL51-R56), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-34bd987125c20aa08a7cd7547b3e3f2bcfb581fd07ad4bbad8dcebe742da4acbL51-R56))
* Add a new tab for exams in the calendar page ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-523cbfe2fd6aeaaadf4887be6f0adfc43dbc03c3a9e078617236fa508098fca7L29-R29), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-0b854772c3dc89a4935981b764d7c85e5e8d0a23e87a952fda38f0d76707341fR78-R80), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-0b854772c3dc89a4935981b764d7c85e5e8d0a23e87a952fda38f0d76707341fL108-R111), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-78074d45bf19cf048805f5e665f5939a10d48ed2a38ffc2c3123d580f4cd4a2dL63-R64), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-ec022af9026981fa46167cdc8ca79c0a398891e72057de40ebdf22884d00e495L63-R64))
  * Use the `t` function to translate the `calendar.exam` key for the name of the exam events in the `CalendarCard` component ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-523cbfe2fd6aeaaadf4887be6f0adfc43dbc03c3a9e078617236fa508098fca7L29-R29))
  * Extract the `focus` query parameter from the `router` object in the `calendar` page and assign the `page` variable to 1 if the `focus` is `exams`, or 0 otherwise ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-0b854772c3dc89a4935981b764d7c85e5e8d0a23e87a952fda38f0d76707341fR78-R80))
  * Pass the `defaultPage` prop to the `SwipeableTabs` component in the `calendar` page with the value of the `page` variable to show the appropriate tab by default based on the `focus` query parameter ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-0b854772c3dc89a4935981b764d7c85e5e8d0a23e87a952fda38f0d76707341fL108-R111))
  * Add the `exams` key and value to the German and English translation files for the `common` namespace to provide the translation for the tab title ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-78074d45bf19cf048805f5e665f5939a10d48ed2a38ffc2c3123d580f4cd4a2dL63-R64), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-ec022af9026981fa46167cdc8ca79c0a398891e72057de40ebdf22884d00e495L63-R64))
* Enable the `SwipeableTabs` component to react to changes in the `defaultPage` prop ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-6a04ff4d4e23225278333eda51a6c83cab2b4b369e9ad1a0632a924f11a990cbL1-R1), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-6a04ff4d4e23225278333eda51a6c83cab2b4b369e9ad1a0632a924f11a990cbL8-R19), [link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-6a04ff4d4e23225278333eda51a6c83cab2b4b369e9ad1a0632a924f11a990cbR39))
  * Import the `useEffect` hook from React in `SwipeableTabs.js` ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-6a04ff4d4e23225278333eda51a6c83cab2b4b369e9ad1a0632a924f11a990cbL1-R1))
  * Add the `defaultPage` prop to the `SwipeableTabs` component to allow the parent component to specify which tab should be shown by default ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-6a04ff4d4e23225278333eda51a6c83cab2b4b369e9ad1a0632a924f11a990cbL8-R19))
  * Add the `defaultPage` prop to the prop types validation of the `SwipeableTabs` component to ensure type safety and documentation ([link](https://github.com/neuland-ingolstadt/neuland.app/pull/306/files?diff=unified&w=0#diff-6a04ff4d4e23225278333eda51a6c83cab2b4b369e9ad1a0632a924f11a990cbR39))

### CODE REVIEW

**Review by M4GNV5 — APPROVED** [review]
image

File: rogue-thi-app/components/cards/CalendarCard.js [hunk 1]
```diff
@@ -26,7 +26,7 @@ export default function CalendarCard () {
       try {
         exams = (await loadExamList())
           .filter(x => !!x.date) // remove exams without a date
-          .map(x => ({ name: `Prüfung ${x.titel}`, begin: x.date }))
+          .map(x => ({ name: `${t('calendar.exam')} ${x.titel}`, begin: x.date }))
```
Comment (Robert27) [inline rogue-thi-app/components/cards/CalendarCard.js:29]: Please add the additional parameter `t`` to the React hook in line 46. [router, t]

File: rogue-thi-app/components/cards/ExamsCard.js [hunk 2]
```diff
@@ -0,0 +1,79 @@
...
+
+import { faUserGraduate } from '@fortawesome/free-solid-svg-icons'
+
+import { loadExamList } from '../../lib/backend-utils/calendar-utils'
+import BaseCard from './BaseCard'
+import { NoSessionError } from '../../lib/backend/thi-session-handler'
+import { formatFriendlyRelativeTime } from '../../lib/date-utils'
+import { useTime } from '../../lib/hooks/time-hook'
+
+import { i18n, useTranslation } from 'next-i18next'
```
Comment (Robert27) [inline rogue-thi-app/components/cards/ExamsCard.js:13]: 'i18n' is defined but never used

**Review by Robert27 — APPROVED** [review]

**Comment by BuildmodeOne** [comment]
> examcards

I love it 😂

**Pull request merged**